### ISSUE TITLE
Cannot create (slack) grafana_contact_point - error 400 type should not be empty

### ISSUE THREAD

**Issue by wiardvanrij** [grafana/terraform-provider-grafana#630]
### Terraform Version
* Terraform v1.1.7
* Terraform Grafana Provider: 1.28.0
* Grafana: Enterprise v9.1.2 

### Affected Resource(s)
Please list the resources as a list, for example:
* grafana_contact_point

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl
resource "grafana_contact_point" "testing" {
    name = "testing"
    slack {
        url = var.slack_webhook
        recipient = "testing-channel-abc"
    }
}

```

### Debug Output
Contains too much information I cannot share.

### Panic Output
```
grafana_contact_point.testing: Creating...
╷
│ Error: status: 400, body: {"message":"invalid object specification: type should not be an empty string","traceID":"00000000000000000000000000000000"}
│
│   with grafana_contact_point.testing,
│   on alert_contact_points.tf line 1, in resource "grafana_contact_point" "testing":
│    1: resource "grafana_contact_point" "testing" {
│
╵
```

### Expected Behavior
Endpoint is created

### Actual Behavior
Grafana throws an error 400: "invalid object specification: type should not be an empty"

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Use my config as described above
2. `terraform apply`

### Important Factoids
Runs on K8s with Istio

### Other stuff ###
- Also tried it against Grafana 9.x version.
- I've checked, but `type` is not a field I can add, it should be inferred from the list.

**Comment by alexweav** [comment]
I tested the following file against a locally ran Grafana 9.1.2 and I did not see any issues:

```
terraform {
  required_version = ">=0.13.0"

  required_providers{
    grafana = {
        source = "grafana/grafana"
        version = "1.28.0"
    }
  }
}

provider "grafana" {
    url = "http://127.0.0.1:3000"
    auth = "admin:admin"
}

resource "grafana_contact_point" "testing" {
    name = "testing"
    slack {
        url = "http://slack"
        recipient = "testing-channel-abc"
    }
}
```

I also tried posting many similar contact points, of mixed type, same type, same name, etc, to hopefully cover any edge cases or parallelism, but did not see any failures.

Are there other resources being provisioned in the same run as the one that's failing? If you delete this one contact point, does everything else work? Does it reproduce if you [disable parallelism](https://www.terraform.io/cli/commands/apply#apply-options) with `-parallelism=1`?

**Comment by alexweav** [comment]
One more useful datapoint - were there any errors in the Grafana logs that correspond to the failed run? Any errors, or even the log entry for when Grafana responded with the 400 would be great additional context.

**Comment by techgnosis** [comment]
> I tested the following file against a locally ran Grafana 9.1.2 and I did not see any issues:
> 
> ```
> terraform {
>   required_version = ">=0.13.0"
> 
>   required_providers{
>     grafana = {
>         source = "grafana/grafana"
>         version = "1.28.0"
>     }
>   }
> }
> 
> provider "grafana" {
>     url = "http://127.0.0.1:3000"
>     auth = "admin:admin"
> }
> 
> resource "grafana_contact_point" "testing" {
>     name = "testing"
>     slack {
>         url = "http://slack"
>         recipient = "testing-channel-abc"
>     }
> }
> ```
> 
> I also tried posting many similar contact points, of mixed type, same type, same name, etc, to hopefully cover any edge cases or parallelism, but did not see any failures.
> 
> Are there other resources being provisioned in the same run as the one that's failing? If you delete this one contact point, does everything else work? Does it reproduce if you [disable parallelism](https://www.terraform.io/cli/commands/apply#apply-options) with `-parallelism=1`?

This worked for me as well. Grafana 9.1.0, Terraform 1.2.1

**Comment by wiardvanrij** [comment]
I will add more details soon. We do run more things but there is only one change. Output:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # grafana_contact_point.testing will be created
  + resource "grafana_contact_point" "testing" {
      + id   = (known after apply)
      + name = "testing"

      + slack {
          + disable_resolve_message = false
          + recipient               = "testing-channel-abc"
          + uid                     = (known after apply)
          + url                     = (sensitive value)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

grafana_contact_point.testing: Creating...
╷
│ Error: status: 400, body: {"message":"invalid object specification: type should not be an empty string","traceID":"00000000000000000000000000000000"}
│
│   with grafana_contact_point.testing,
│   on alert_contact_points.tf line 1, in resource "grafana_contact_point" "testing":
│    1: resource "grafana_contact_point" "testing" {
│
╵
Releasing state lock. This may take a few moments...
```

`-parallelism=1` did not make a difference. 

Will update more later!

**Comment by alexweav** [comment]
Hmm, very interesting :thinking:. I tried a different test case - I took an existing environment with lots of different resources already managed by Terraform. I then added this contact point to it, to try and simulate this setup a bit closer, or see if there was an issue with having state already on the server.

My TF output is character-for-character identical to yours, but instead it doesn't error:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # grafana_contact_point.testing will be created
  + resource "grafana_contact_point" "testing" {
      + id   = (known after apply)
      + name = "testing"

      + slack {
          + disable_resolve_message = false
          + recipient               = "testing-channel-abc"
          + uid                     = (known after apply)
          + url                     = (sensitive value)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

grafana_contact_point.testing: Creating...
grafana_contact_point.testing: Creation complete after 0s [id=XgJF4CW4z]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

We are definitely using the same version of the provider, too - so I wonder if the issue is instead on the side of the server? The API that Terraform is talking to is rather new. We have fixed a few bugs in it recently, but it seems a possible candidate. This is also where Grafana logs would help us dig a little further.

What other contact points might there already be on the server? Are you able to hand-create the exact same contact point in the UI, or do you only get the failure through Terraform?

**Comment by wiardvanrij** [comment]
```
logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-08-31T18:05:26.828931796Z level=info msg="Request Completed" method=POST path=/api/ma/events status=201 remote_addr=xxxx time_ms=7 duration=7.870116ms size=4 referer="https://our-url.com/alerting/notifications/receivers/new?alertmanager=grafana" traceID=00000000000000000000000000000000
logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-08-31T18:06:13.05181687Z level=info msg="Request Completed" method=POST path=/api/alertmanager/grafana/config/api/v1/alerts status=202 remote_addr=xxxx time_ms=48 duration=48.086116ms size=35 referer="https://our-url.com/alerting/notifications/receivers/new?alertmanager=grafana" traceID=00000000000000000000000000000000
logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-08-31T18:07:07.453296266Z level=info msg="Request Completed" method=POST path=/api/alertmanager/grafana/config/api/v1/alerts status=202 remote_addr=xxxx time_ms=25 duration=25.942662ms size=35 referer=https://our-url.com/alerting/notifications traceID=00000000000000000000000000000000
logger=context traceID=00000000000000000000000000000000 userId=0 orgId=1 uname= t=2022-08-31T18:09:37.070755642Z level=error msg="Error 1051: Unknown table 'T'" error="Error 1051: Unknown table 'T'" remote_addr=xxxx traceID=00000000000000000000000000000000
logger=context traceID=00000000000000000000000000000000 userId=0 orgId=1 uname= t=2022-08-31T18:09:37.070809993Z level=error msg="Request Completed" method=POST path=/api/v1/provisioning/contact-points status=500 remote_addr=xxxx time_ms=31 duration=31.478513ms size=88 referer= traceID=00000000000000000000000000000000
logger=context traceID=00000000000000000000000000000000 userId=0 orgId=1 uname= t=2022-08-31T18:09:42.776102466Z level=error msg="invalid object specification: type should not be an empty string" error="invalid object specification: type should not be an empty string" remote_addr=xxxx traceID=00000000000000000000000000000000
logger=context traceID=00000000000000000000000000000000 userId=0 orgId=1 uname= t=2022-08-31T18:09:42.776145481Z level=info msg="Request Completed" method=POST path=/api/v1/provisioning/contact-points status=400 remote_addr=xxxx time_ms=41 duration=41.576263ms size=123 referer= traceID=00000000000000000000000000000000
```

I do get an error before it. I'm also looking into it now. I can create it via the UI. Be it that you cannot add the Recipient when you define a webhook.. but that's a different issue.

**Comment by wiardvanrij** [comment]
Maybe useful to note is that we are using RDS. Just enabled the sql logs and this is the query before the error, which looks (IMO) a bit weird:

```
logger=sqlstore.xorm t=2022-08-31T18:36:09.511899759Z level=info msg="[SQL] \n\t\tINSERT INTO alert_configuration\n\t\t(alertmanager_configuration, configuration_hash, configuration_version, org_id, created_at, `default`) \n\t\tSELECT T.* FROM (SELECT ? AS alertmanager_configuration,? AS configuration_hash,? AS configuration_version,? AS org_id,? AS created_at,? AS 'default') AS T\n\t\tWHERE\n\t\tEXISTS (\n\t\t\tSELECT 1 \n\t\t\tFROM alert_configuration \n\t\t\tWHERE \n\t\t\t\torg_id = ? \n\t\t\tAND \n\t\t\t\tid = (SELECT MAX(id) FROM alert_configuration WHERE org_id = ?) \n\t\t\tAND \n\t\t\t\tconfiguration_hash = ?\n\t\t) []interface {}{\"{\\\"template_files\\\":{},\\\"alertmanager_config\\\":{\\\"route\\\":{\\\"receiver\\\":\\\"grafana-default-email\\\",\\\"routes\\\":[{\\\"receiver\\\":\\\"grafana-default-email\\\"}]},\\\"templates\\\":null,\\\"receivers\\\":[{\\\"name\\\":\\\"grafana-default-email\\\",\\\"grafana_managed_receiver_configs\\\":[{\\\"uid\\\":\\\"bAKbDjZ4z\\\",\\\"name\\\":\\\"email receiver\\\",\\\"type\\\":\\\"email\\\",\\\"disableResolveMessage\\\":false,\\\"settings\\\":{\\\"addresses\\\":\\\"\\\\u003cexample@email.com\\\\u003e\\\"},\\\"secureSettings\\\":null}]},{\\\"name\\\":\\\"testing\\\",\\\"grafana_managed_receiver_configs\\\":[{\\\"uid\\\":\\\"nCnvKCZVz\\\",\\\"name\\\":\\\"testing\\\",\\\"type\\\":\\\"slack\\\",\\\"disableResolveMessage\\\":false,\\\"settings\\\":{\\\"endpointUrl\\\":\\\"\\\",\\\"icon_emoji\\\":\\\"\\\",\\\"icon_url\\\":\\\"\\\",\\\"mentionChannel\\\":\\\"\\\",\\\"mentionGroups\\\":\\\"\\\",\\\"mentionUsers\\\":\\\"\\\",\\\"recipient\\\":\\\"\\\",\\\"text\\\":\\\"\\\",\\\"title\\\":\\\"\\\",\\\"username\\\":\\\"\\\"},\\\"secureSettings\\\":{\\\"token\\\":\\\"redacted=\\\",\\\"url\\\":\\\"fooooooooooooooooooooo-redacted=\\\"}}]}]}}\", \"redacated\", \"v1\", 1, 1661970969, \"0\", 1, 1, \"1c2815c03601c302f72854fc0572d9d1\"} - took: 943.382µs" caller="[log.go:187 logger.go:63 session_raw.go:165 session_raw.go:180 session_raw.go:228 alertmanager.go:106 transactions.go:49 transactions.go:20 alertmanager.go:97 contactpoints.go:192 transactions.go:30 transactions.go:49 transactions.go:28 transactions.go:24 transactions.go:6 contactpoints.go:191 api_provisioning.go:111 provisioning.go:32 generated_base_api_provisioning.go:113 ngalert.go:315 web_hack.go:39 server.go:2047 context.go:60 context.go:55 ratelimiting.go:53 web_hack.go:31 server.go:2047 context.go:60 context.go:55 query_middleware.go:353 web_hack.go:51 server.go:2047 context.go:60 context.go:55 macaron.go:139 server.go:2047 csp.go:22 server.go:2047 context.go:60 context.go:55 macaron.go:139 server.go:2047 render.go:44 server.go:2047 context.go:60 context.go:55 macaron.go:139 server.go:2047 csrf.go:64 server.go:2047 context.go:60 context.go:55 recovery.go:168 web_hack.go:51 server.go:2047 context.go:60 context.go:55 logger.go:35 web_hack.go:31 server.go:2047 context.go:60 context.go:55 request_metrics.go:57 web_hack.go:31 server.go:2047 context.go:60 context.go:55 request_tracing.go:79 web_hack.go:31 server.go:2047 context.go:60 router.go:156 router.go:213 macaron.go:170]"
```

Main thing is, the log tells me it cannot find table T which should be an 'alias' for: ```SELECT T.* FROM (SELECT ? AS alertmanager_configuration,? AS configuration_hash,? AS configuration_version,? AS org_id,? AS created_at,? AS 'default') AS T``` 

Hope this somewhat helps? - Also not sure if this is exactly why it does not work.

**Comment by alexweav** [comment]
Ah nice, this was great information. This is 100% the cause, I was able to reproduce it by intentionally inducing a DB-layer failure.

Looks like Grafana's MySQL backend isn't quite compatible with that RDS engine. Which engine are you using? I will file an issue to the Grafana project.

This is a surprising one, as it does work against MySQL interfaces on other cloud providers. Perhaps there is a versioning difference..

**Comment by alexweav** [comment]
So, this caused the server to return a 500. You can see this first request in the log.

This then exposed a second bug deep in the [golang-client](https://github.com/grafana/grafana-api-golang-client) project, underneath this provider. It has [retry](https://github.com/grafana/grafana-api-golang-client) logic for failed requests. When a retry happens, it tries to read an `io.Reader` containing the request body twice!

The second read returns an empty slice of bytes, which is then marshalled into an empty Contact Point struct, which has... an empty string for `Type`! Hence, `type should not be an empty string`.

This bug has been around for about a year now, interesting that we only just now found it.

**Comment by jdgeisler** [comment]
@alexweav following up for @wiardvanrij - the RDS engine version we are using is `5.7.mysql_aurora.2.07.2`.

Thanks!

**Comment by wiardvanrij** [comment]
Thanks Alexander. I would have not figured that out. Awesome.

As for workaround / timeline. Would you reckon this takes a while to get fixed due to the more deeper/nested nature of the issue? Because I'm otherwise open to pick a different DB solution if that unblocks us. Ideally I would like to get this up and running somewhere next week. Even if that means putting some effort into building a different DB stack

**Comment by alexweav** [comment]
Core grafana issue on this: https://github.com/grafana/grafana/issues/54556

As far as I can tell, our docs are very vague as to which flavors of "database-compatible" cloud tools we actually support. This page simply says `MySQL` and nothing else... https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#database

In practice, we don't actively test Grafana on every "MySQL-compatible" database product on every cloud provider... however I can confirm that both native MySQL and GCP Cloud SQL w/MySQL interface work correctly here, as we use both internally often. I'm sharing the issue above around in order to get a more definitive answer on this.

**Comment by alexweav** [comment]
PR for the retry issue: https://github.com/grafana/grafana-api-golang-client/pull/109

After applying this change locally and testing against an affected server, it now returns the right error message. We don't get `400 - ... type should not be an empty string`, we now get `500 - no such table: ...` as we'd expect.

**Comment by wiardvanrij** [comment]
Created a case at AWS as well. Will provide updates whenever I got them.

**Comment by joeblubaugh** [comment]
The Grafana-supported databases are listed here: https://grafana.com/docs/grafana/latest/setup-grafana/installation/#supported-databases

RDS/MySQL isn't officially supported, but we can look into whether re-writing this query would be enough to avoid the issue.

**Comment by subkryl** [comment]
I have the same problem with the helm stack from "https://grafana.github.io/helm-charts" grafana/grafana.

It's grafana 9.0.2, I guess they use sqlite db and no mysql like database. Maybe there is a relation ?

The version of the helm chart is : 6.32.2

return from TF : Error: status: 400, body: {"message":"invalid object specification: type should not be an empty string","traceID":"00000000000000000000000000000000"}

I use login/password to auth with the grafana provider.

And from grafana logs:

logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-09-08T20:50:42.73898801Z level=error msg="no secrets configured for type 'prometheus-alertmanager'" error="no secrets configured for type 'prometheus-alertmanager'" remote_addr=127.0.0.1 traceID=00000000000000000000000000000000

logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-09-08T20:50:42.739117323Z level=error msg="Request Completed" method=POST path=/api/v1/provisioning/contact-points status=500 remote_addr=127.0.0.1 time_ms=41 duration=41.445602ms size=115 referer= traceID=00000000000000000000000000000000

logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-09-08T20:50:47.803110662Z level=error msg="invalid object specification: type should not be an empty string" error="invalid object specification: type should not be an empty string" remote_addr=127.0.0.1 traceID=00000000000000000000000000000000

logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-09-08T20:50:47.803203505Z level=info msg="Request Completed" method=POST path=/api/v1/provisioning/contact-points status=400 remote_addr=127.0.0.1 time_ms=40 duration=40.988882ms size=123 referer= traceID=00000000000000000000000000000000

**Comment by subkryl** [comment]
> Note, it works with "slack" object like:

resource "grafana_contact_point" "testing" {
  name = "testing"
  slack {
    url = "http://slack"
    recipient = "testing-channel-abc"
  }
}

But not with "alertmanager" object:

resource "grafana_contact_point" "alertmanager" {
  name = "alertmanager"
  alertmanager {
    url = var.ALERTMANAGER_URL
  }
}

**Comment by alexweav** [comment]
@subkryl Try updating to the latest provider. What error do you see then?

See the PR linked just above your comment for more context, but it's basically giving you the wrong error message. The real error was hidden by some faulty retry logic. This should be fixed in the latest release.

It won't make the error go away outright, but it will tell you what actually went wrong, instead of the red herring "type should not be empty."

**Comment by alexweav** [comment]
I believe we'll close this one out for now, as the above PR is submitted and released. We should not see this particular error message any more in the latest versions.

The underlying error in @wiardvanrij's case now has a separate issue tracking it in grafana/grafana, so that should be the place for discussion on RDS going forward. Please let us know if anyone runs into this again post-1.28.1 and we can re-open.
